File: 백준/Gold/14502. 연구소/연구소.cc
#include<iostream>
#include<vector>
#include<queue>

using namespace std;

int a, b;
int arr[8][8];
vector<pair<int, int>> virus;
int MAX;
int dx[4] = { 1,-1,0,0 };
int dy[4] = { 0,0,1,-1 };

void search() {
	
	int cnt = a * b;
	int wall = 0;

	//바이러스 계산용 테이블
	int temp[8][8] = { 0, };

	//복사
	for (int i = 0; i < a; i++) {
		for (int j = 0; j < b; j++) {
			//구역 복사
			temp[i][j] = arr[i][j];

			//벽 개수
			if (arr[i][j] == 1) {
				wall++;
			}
		}
	}

	//벽 개수만큼 안전구역 제거
	cnt -= wall;

	//BFS용 큐
	queue<pair<int, int>>BFS;
	for (int n = 0; n < virus.size(); n++) {
		//virus 시작 지점 큐에 삽입
		BFS.push(virus[n]);

		
		//위험구역 개수
		cnt--;
		
	}

	//BFS 탐색 시작
	while (!BFS.empty()) {
		int x = BFS.front().first;
		int y = BFS.front().second;

		BFS.pop();

		//상하좌우 탐색
		for (int k = 0; k < 4; k++) {
			int tx = x + dx[k];
			int ty = y + dy[k];
			//이동 가능한 경우
			if (tx >= 0 && ty >= 0 && tx < a && ty < b) {
				//탐색 가능한 경우
				if (temp[tx][ty] == 0) {
					temp[tx][ty] = 2;
					cnt--;
					BFS.push({ tx,ty });
				}
			}
		}
	}
	
	int tem = 0;
	for (int i = 0; i < a; i++) {
		for (int j = 0; j < b; j++) {
			if (temp[i][j] == 0)tem++;
		}
	}

	MAX = (MAX < cnt ? cnt : MAX);
}

void select(int x,int y, int N) {

	//전체 방 탐색
	for (int i = x; i < a; i++) {
		//첫 번째 반복일 경우 y 아닐 경우 0열부터 탐색
		for (int j = (i == x ? y : 0); j < b; j++) {
			//빈 칸인 경우
			if (arr[i][j] == 0) {
				//벽 세우기
				arr[i][j] = 1;

				//마지막 벽이라면
				if (N == 3) {
					search();
				}
				//마지막 벽이 아니라면
				else {
					select(i, j, N + 1);
				}
				
				//벽 다시 돌려놓기
				arr[i][j] = 0;
			}
		}
	}
}

int main() {

	cin >> a >> b;

	MAX = 0;

	//방 구조 입력
	for (int i = 0; i < a; i++) {
		for (int j = 0; j < b; j++) {
			cin >> arr[i][j];
			//바이러스인 경우 위치 저장
			if (arr[i][j] == 2) {
				virus.push_back({ i,j });
			}
		}
	}

	//벽 선택
	select(0, 0, 1);

	//최대 안전구역 출력
	cout << MAX << "\n";

	return 0;
}

File: 백준/Gold/14502. 연구소/연구소.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder sb = new StringBuilder();
	static StringTokenizer st;

	static int N, M;
	static int map[][];
	static int ans = 0;

	static class Pair {
		int x, y;

		public Pair(int x, int y) {
			super();
			this.x = x;
			this.y = y;
		}

	}

	static void select(int cnt) {
		if (cnt == 3) {
			boolean visit[][] = new boolean[N][M];

			Queue<Pair> queue = new LinkedList<>();

			int dx[] = { 1, -1, 0, 0 };
			int dy[] = { 0, 0, 1, -1 };

			int temp[][] = new int[N][M];
			for (int i = 0; i < N; i++) {
				for (int j = 0; j < M; j++) {
					temp[i][j] = map[i][j];
					if (temp[i][j] == 2) {
						visit[i][j] = true;
						queue.add(new Pair(i, j));
					}
				}
			}

			while (!queue.isEmpty()) {
				Pair p = queue.poll();

				for (int k = 0; k < 4; k++) {
					int tx = p.x + dx[k];
					int ty = p.y + dy[k];

					if (tx >= 0 && tx < N && ty >= 0 && ty < M) {
						if (!visit[tx][ty] && temp[tx][ty] == 0) {
							visit[tx][ty] = true;
							temp[tx][ty] = 2;
							queue.add(new Pair(tx, ty));
						}
					}
				}
			}

			int tempCnt = 0;
			for (int i = 0; i < N; i++) {
				for (int j = 0; j < M; j++) {
					if (temp[i][j] == 0)
						tempCnt++;
				}
			}

			ans = Math.max(ans, tempCnt);

			return;
		}

		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				if (map[i][j] == 0) {
					map[i][j] = 1;
					select(cnt + 1);
					map[i][j] = 0;
				}
			}
		}
	}

	public static void main(String[] args) throws NumberFormatException, IOException {

		st = new StringTokenizer(in.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());

		map = new int[N][M];
		
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(in.readLine());
			for (int j = 0; j < M; j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		select(0);

		System.out.println(ans);
	}
}